runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/io/SdkByteReadChannelJVM.kt

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import kotlinx.coroutines.runBlocking
+import java.io.InputStream
+import java.util.*
+
+/**
+ * Create a blocking [InputStream] that blocks everytime the channel suspends at [SdkByteReadChannel.read]
+ */
+public fun SdkByteReadChannel.toInputStream(): InputStream = InputAdapter(this)
+
+private const val DEFAULT_READ_BYTES = 8192L
+private class InputAdapter(private val ch: SdkByteReadChannel) : InputStream() {
```

Comment:
**Question**: This adapter works via `runBlocking` on every `read`. That will work obvi but seems like it may be slow. Would it be better to spin up a background coroutine to handle reads that lives for the lifetime of the adapter?

Reply:
`InputStream.read` is a blocking call. `runBlocking` is designed to bridge blocking and non-blocking worlds (it does so by blocking the current thread). Spinning up a coroutine will result in tying up two threads and/or more context switches to accomplish the same thing.

Reply:
As far as I understand (could be wrong, ofc) [`runBlocking` _starts a new coroutine_ and blocks the current thread](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/run-blocking.html#:~:text=Runs%20a%20new%20coroutine%20and%20blocks%20the%20current%20thread%20until%20its%20completion). Doing that on every 8KB chunk _sounds_ slower than starting a new coroutine _once_. Sure, it'll still block on every `read` call but the cost of setting up a new coroutine was already paid.

Reply:
TL;DR I'm not at all worried about the overhead of `runBlocking` or starting a new coroutine.

----
`runBlocking` does start a new coroutine and blocks the current thread. This is fine though, coroutines are extremely lightweight. The only benchmarking I can find is [here](https://stackoverflow.com/questions/50230466/kotlin-withcontext-vs-async-await/50231191#50231191) which would suggest **140** byte allocation and **100 nanoseconds**. 

Introducing a background coroutine will likely introduce additional context switches and a more complicated implementation to synchronize the coroutine and blocking `read` call. It also needs a `CoroutineScope` to launch into which would also mean taking an additional parameter and dealing with cancellation, error propagation, etc.